engine/src/errors.rs

```diff
@@ -0,0 +1,42 @@
+use thiserror::Error;
+use wasmtime::MemoryAccessError;
+
+#[derive(Error, Debug)]
+pub enum ServalEngineError {
+    #[error("Failed to get the default export of the binary")]
+    DefaultExportUnavailable,
+
+    #[error("Error initializing engine: {0}")]
+    EngineInitializationError(anyhow::Error),
+
+    #[error("Error executing binary")]
+    ExecutionError {
+        stdout: Vec<u8>,
+        stderr: Vec<u8>,
+        error: anyhow::Error,
+    },
+
+    #[error("The binary's default export does not match the expected function signature")]
+    InvalidDefaultExportFunctionSignature,
+
+    #[error("Guest alloc function did not return a valid pointer")]
+    InteropAllocFailed,
+
+    #[error("Guest does not export an alloc function")]
+    InteropAllocUnavailable,
+
+    #[error("Reading or writing from guest memory failed")]
+    InteropMemoryAccessError(MemoryAccessError),
+
+    #[error("std::io::Error: {0}")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Failed to load WASM module")]
+    ModuleLoadError(anyhow::Error),
+
+    #[error("Error reading bytes from stderr pipe")]
```

Comment:
One could argue that these could be collapsed into a single `StandardIOReadError`, since there's not really any actionable decision you could make in either case.

I am open to feedback on any of these error variants.